### PR TITLE
VCST-1982: Export sitemap to store assets

### PR DESCRIPTION
**Pull request by OlegoO**
## Description

feat: This update introduces the ability to export sitemaps to store assets. There are two supported options: manual export from the back office and automated export via a scheduled job.

* Manual Export: Can be initiated from the back office.
* Automated Export: This is configurable through a scheduled job. By default, the automated export is disabled. To enable it, you must first activate it globally and then at the store level. The job schedule can be configured using a cron expression, with the default setting being daily at 00:00.

**Global Settings:**
1. Enable export sitemap files by job - this check indicates whether the export to assets store assets folder should be enabled. Save all sitemap artefacts to stores/{storeId}/sitemap asset folder. By default, false.
2. Export to store assets cron expression - the cron expression for export sitemap files to store assets folder by schedule. By default, '0 0 * * *' daily at 00:00.

**Store Settings:**
1. Enable export sitemap files by job - this option allows you to export sitemap files for a store. By default, it is false. Global Settings should be enabled.

**Permissions:**
1. sitemaps:exportToStoreAssets - allows to control ability to run export to store assets for the a role.
  
Screenshots:

image

image

image

image

### Jira-link:

https://virtocommerce.atlassian.net/browse/VCST-1982
### Artifact URL:
https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.Sitemaps_3.810.0-pr-77-be61.zip

### CODE REVIEW

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    protected virtual async Task ProcessStore(Store store)
+    {
+        ArgumentNullException.ThrowIfNull(store);
+
+        var isEnabled = store.Settings.GetValue<bool>(Core.ModuleConstants.Settings.General.EnableExportToAssetsJob);
+        if (isEnabled)
+        {
+            var outputAssetFolder = string.Format(Core.ModuleConstants.StoreAssetsOutputFolderTemplate, store.Id);
+
+            _logger.LogInformation("Starting export {sitemapUrl} for store {storeId} to {outputAssetFolder}.", Core.ModuleConstants.SitemapXmlFileName, store.Id, outputAssetFolder); // Log success
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:67]: {sitemapUrl} sounds strange for SitemapXmlFileName variable. Maybe it's better to use {sitemapFileName}?

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            foreach (var sitemapUrl in await _sitemapXmlGenerator.GetSitemapUrlsAsync(store.Id, store.Url))
+            {
+                _logger.LogInformation("Starting export {sitemapUrl} for store {storeId} to {outputAssetFolder}.", sitemapUrl, store.Id, outputAssetFolder); // Log success
+
+                await ExportSitemapPartAsync(store, outputAssetFolder, sitemapUrl);
+            }
+        }
+    }
+
+
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:80]: An extra empty line

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    protected virtual async Task ExportSitemapPartAsync(Store store, string outputAssetFolder, string sitemapUrl)
+    {
+        var relativeUrl = $"{outputAssetFolder.Trim('/')}/{sitemapUrl.Trim('/')}";
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:83]: If we work with real url link we may use Uri class for better experience and management.

var uri = new Uri(new Uri(outputAssetFolde), sitemapUrl);
 _blobStorageProvider.OpenWrite(uri.AbsoluteUri);

This approach will always give a correctly formed link, no matter how many "/" characters there would be in the original parts.

File: src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs [hunk 4]
```diff
@@ -412,11 +477,20 @@ void SendNotificationWithProgressInfo(ExportImportProgressInfo c)
...
-                await stream.CopyToAsync(sitemapPartStream);
-            }
+            using var sitemapPartStream = sitemapPart.Open();
+            using var stream = await _sitemapXmlGenerator.GenerateSitemapXmlAsync(storeId, baseUrl, sitemapUrl, progressCallback);
+            await stream.CopyToAsync(sitemapPartStream);
+        }
+
+        private async Task<BlobInfo> ExportSitemapPartAsync(string storeId, string storeUrl, string outputAssetFolder, string sitemapUrl, Action<ExportImportProgressInfo> progressCallback)
+        {
+            var relativeUrl = $"{outputAssetFolder.Trim('/')}/{sitemapUrl.Trim('/')}";
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs:487]: The same as in SitemapProcessJob.cs - if we work with real url link we may use Uri class for better experience and management.
Secondly, we basically have the same code for this function in two places (here and in SitemapProcessJob)

**Review by Ljutyj — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.SitemapsModule.Web/Model/PushNotifications/SitemapExportToAssetNotification.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+using VirtoCommerce.Platform.Core.PushNotifications;
+
+namespace VirtoCommerce.SitemapsModule.Web.Model.PushNotifications
+{
+    /// <summary>
+    ///
```
[the remark this answers is not in the document]
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Web/Model/PushNotifications/SitemapExportToAssetNotification.cs:9]: Fixed

File: src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs [hunk 6]
```diff
@@ -329,21 +326,83 @@ public async Task<ActionResult<SitemapDownloadNotification>> DownloadSitemap(str
...
+
+            await _notifier.SendAsync(notification);
+
+            BackgroundJob.Enqueue(() => BackgroundExportToAssets(storeId, baseUrl, notification));
+
+            return Ok(notification);
+        }
+
+        [ApiExplorerSettings(IgnoreApi = true)]
+        public Task BackgroundExportToAssets(string storeId, string baseUrl, SitemapExportToAssetNotification notification)
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs:358]: Do we really need this public method that does nothing but call a private method (except for checking the storeId)? Does it really have to be public? Maybe we can get rid of it and use only the private method? We also may convert the private method to public if necessary.

**Review by Ljutyj — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs [hunk 7]
```diff
@@ -412,11 +477,20 @@ void SendNotificationWithProgressInfo(ExportImportProgressInfo c)
...
-            {
-                var stream = await _sitemapXmlGenerator.GenerateSitemapXmlAsync(storeId, baseUrl, sitemapUrl, progressCallback);
-                await stream.CopyToAsync(sitemapPartStream);
-            }
+            using var sitemapPartStream = sitemapPart.Open();
+            using var stream = await _sitemapXmlGenerator.GenerateSitemapXmlAsync(storeId, baseUrl, sitemapUrl, progressCallback);
+            await stream.CopyToAsync(sitemapPartStream);
+        }
+
+        private async Task<BlobInfo> ExportSitemapPartAsync(string storeId, string storeUrl, string outputAssetFolder, string sitemapUrl, Action<ExportImportProgressInfo> progressCallback)
```
[the remark this answers is not in the document]
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs]: Fixed

File: src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs (hunk 6 above)
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Web/Controllers/Api/SitemapsModuleApiController.cs]: Fixed. Move all job login to job service.

File: src/VirtoCommerce.SitemapsModule.Data/Extensions/RelativePathUtils.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
+using System;
+
+namespace VirtoCommerce.SitemapsModule.Data.Extensions;
+
+public static class RelativePathUtils
+{
+    public static string Combine(string parentRelativePath, string childRelativePath)
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Data/Extensions/RelativePathUtils.cs:7]: I've only found one place where it's used now. There used to be two. Is there any sense of a separate existence now?))

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        finally
+        {
+            notification.Finished = DateTime.UtcNow;
+            await _notifier.SendAsync(notification);
+        }
+    }
+
+    private async Task InnerBackgroundDownload(string storeId, string baseUrl, string localTmpFolder, SitemapDownloadNotification notification)
+    {
+        void SendNotificationWithProgressInfo(ExportImportProgressInfo c)
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:136]: InnerBackgroundExportToAssets method has the same local function.  What if we make a method that will receive the notification and return this function to us?

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+            _logger.LogInformation("Starting export {sitemapUrl} for store {storeId} to {outputAssetFolder}.", sitemapUrl, store.Id, outputAssetFolder); // Log success
+
+            await ExportSitemapPartAsync(store.Id, store.Url, outputAssetFolder, sitemapUrl, null);
+        }
+    }
+
+    private async Task CreateSitemapPartAsync(ZipArchive zipArchive, string storeId, string baseUrl, string sitemapUrl, Action<ExportImportProgressInfo> progressCallback)
+    {
+        var sitemapPart = zipArchive.CreateEntry(sitemapUrl, CompressionLevel.Optimal);
+        using var sitemapPartStream = sitemapPart.Open();
```
Comment (Ljutyj) [inline src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:220]: I would add an empty line between 219 and 220 lines . The code looks similar to the same in the ExportSitemapPartAsync method, but stylistically designed differently. It attracts attention.

**Review by Ljutyj — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    public Task BackgroundExportToAssets(string storeId, string baseUrl, SitemapExportToAssetNotification notification)
+    {
+        ArgumentNullException.ThrowIfNullOrWhiteSpace(storeId);
+
+        if (!Uri.TryCreate(baseUrl, UriKind.Absolute, out _))
+        {
+            throw new ArgumentException($"Incorrect base URL {baseUrl}");
+        }
```
[the remark this answers is not in the document]
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:91]: let's keep it simple

File: src/VirtoCommerce.SitemapsModule.Data/Extensions/RelativePathUtils.cs (hunk 8 above)
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Data/Extensions/RelativePathUtils.cs:7]: Fixed

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs (hunk 10 above)
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs:201]: Fixed

File: src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs (hunk 9 above)
Comment (OlegoO) [reply src/VirtoCommerce.SitemapsModule.Data/BackgroundJobs/SitemapProcessJob.cs]: Fixed

**Review by Ljutyj — APPROVED** [review]

**Pull request merged**